pal/src/misc/random.cpp

```diff
@@ -18,34 +18,47 @@ Module Name:
 --*/
 
 #include "pal/palinternal.h"
-#include <sodium.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <mutex>
 
-/*++
-Function :
-RANDOMInitialize
-
-Initialize libsodium
 
-(no parameters)
+#define RANDOM_CACHE_SIZE 8
+constexpr int rand_byte_len = sizeof(unsigned int) * RANDOM_CACHE_SIZE;
+static union {
+  char rstack[rand_byte_len];
+  unsigned int result[RANDOM_CACHE_SIZE];
+} random_cache;
+static unsigned cache_index(RANDOM_CACHE_SIZE);
+static std::mutex random_generator_mutex;
 
-Return value :
-TRUE  if libsodium initialization succeeded
-FALSE otherwise
---*/
-BOOL RANDOMInitialize(void)
+static bool GetRandom(unsigned int *result) noexcept
 {
-    if (sodium_init() == -1)
-    {
-        return FALSE;
+    std::lock_guard<std::mutex> guard(random_generator_mutex);
```

Comment:
I wonder if is another optimization you could do here- if you initialize cache_index to -1 in the case where the cache is initialized/reset, then you could skip the mutex enter and instead do an InterlockedIncrement first. If the incremented index (which is what is returned) is greater than RANDOM_CACHE_SIZE, you can then grab the mutex and repopulate the cache- that way, we have a fast path and a slow path (and you enter the mutex only in the slow path). Not sure if it saves us much, just a thought since you are trying to optimize this code anyway